### PR TITLE
fix: pytree - better handling of scalar values 

### PR DESCRIPTION
**Pull request by srossross**
Part of https://github.com/pytorch/functorch/issues/1082 and https://github.com/pytorch/functorch/issues/439

This removes the need for `Some weird edge case requires us to spell out the following see test_out_dims_edge_case`

I found this when trying to add support for None values:

`_broadcast_to_and_flatten([1], *)` should return `[1]` but instead returns `None`

Current behaviour
```py
In [2]: a,b = pytree.tree_flatten(1)
In [3]: c, d = pytree.tree_flatten([1])

In [4]: pytree._broadcast_to_and_flatten(1, b)
Out[4]: [1]

In [5]: pytree._broadcast_to_and_flatten(1, d)
Out[5]: [1]

In [6]: pytree._broadcast_to_and_flatten([1], b)
Out[6]: None
In [7]: pytree._broadcast_to_and_flatten([1], d)
Out[7]: [1]
```
Behaviour after PR:
```py

In [1]: from torch.utils import _pytree  as pytree

In [2]: a,b = pytree.tree_flatten(1)
In [3]: c, d = pytree.tree_flatten([1])

In [4]: pytree._broadcast_to_and_flatten(1, b)
Out[4]: [1]
In [5]: pytree._broadcast_to_and_flatten(1, d)
Out[5]: [1]

In [6]: pytree._broadcast_to_and_flatten([1], b)
Out[6]: [1]
In [7]: pytree._broadcast_to_and_flatten([1], d)
Out[7]: [1]
```

### CODE REVIEW

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -304,7 +304,11 @@ def _broadcast_to_and_flatten(pytree: PyTree, spec: TreeSpec) -> Optional[List[A
     if _is_leaf(pytree):
         return [pytree] * spec.num_leaves
     if isinstance(spec, LeafSpec):
-        return None
+        if isinstance(pytree, (list, tuple)):
```
Comment (srossross) [inline torch/utils/_pytree.py:307]: Should be more general? What if it is a `dict`, `OrderedDict` or `namedtuple` in `SUPPORTED_NODES`

File: torch/_functorch/vmap.py [hunk 2]
```diff
@@ -152,20 +152,9 @@ def incompatible_error():
             f'out_dims has structure {tree_flatten(out_dims)[1]} but outputs '
             f'has structure {output_spec}.')
 
-    if isinstance(batched_outputs, torch.Tensor):
-        # Some weird edge case requires us to spell out the following
-        # see test_out_dims_edge_case
```
Comment (srossross) [inline torch/_functorch/vmap.py:157]: confirmed that `test_out_dims_edge_case` still works

File: torch/_functorch/vmap.py [hunk 3]
```diff
@@ -152,20 +152,9 @@ def incompatible_error():
...
-            out_dims = out_dims[0]
-        else:
-            incompatible_error()
-    else:
-        flat_out_dims = _broadcast_to_and_flatten(out_dims, output_spec)
-        if flat_out_dims is None:
-            incompatible_error()
+    flat_out_dims = _broadcast_to_and_flatten(out_dims, output_spec)
+    if flat_out_dims is None:
+        incompatible_error()
```
Comment (zou3519) [inline torch/_functorch/vmap.py:157]: > `_broadcast_to_and_flatten([1], *)` should return `[1]` but instead returns `None`

What is the argument for why this should be the behavior?

I think `test_out_dims_edge_case` is wrong and we should not be allowing that behavior, but I do not remember why we added the special case for it.

Morally, in `_broadcast_to_and_flatten(a, b)`, I'd expect b's tree-structure to be "larger" than a's tree-structure, and this is not the case with `_broadcast_to_and_flatten([1], *)`

Comment (srossross) [reply torch/_functorch/vmap.py:157]: My goal was to remove the edge case from `torch/_functorch/vmap.py:155`

Comment (zou3519) [reply torch/_functorch/vmap.py:157]: I see, and the edge case is making it difficult to add support for `None` returns?

Directionally we should probably remove the edge case - it looks like JAX does not support it, so we shouldn't either.

In order to remove it, I'd be OK with deprecating it in a release and deleting it in the subsequent release. Open question to if we can just delete the edge case in a single release and label it as BC-breaking. If we want to go down this path, then we should do some code archeology:
1) why was the edge case added in the first place? I don't remember. It's possible that either (a) JAX used to support it and we copied the behavior, or (b) I got confused about what the output of a Python function is.
2) Try to browse usages of functorch.vmap on GitHub (maybe via cs.github.com?). There should be at most ~300 of them right now. See if any do rely on this edge case and if we're breaking anyone.

Off of the top of my head I don't think anyone really uses the out_dims flag, so it might be safe to delete the edge case in a single release.

Comment (srossross) [reply torch/_functorch/vmap.py:157]: > I see, and the edge case is making it difficult to add support for None returns?

It is not 'difficult' I just needed to add more logic to the edge case handling, so I thought I would try to remove it first.
